### ISSUE TITLE
picker.setMargins() not running in Android Phonegap

### ISSUE THREAD

**Issue by linkito** [Scandit/BarcodeScannerPlugin#41]
Hi.

Im trying to use picker.setMargins(new Scandit.Margins(0, 50, 0, 50), null, 0);
but the scanner area is opened in fullscreen always... But is strange, because when you press in the area that supposedly its outside the scanner area, the buttons of this area runs well...

Anybody can help me? Thanks!

My code:

```javascript
settings.setSymbologyEnabled(Scandit.Barcode.Symbology.CODE128, true);
settings.setSymbologyEnabled(Scandit.Barcode.Symbology.UPC12, true);
settings.setSymbologyEnabled(Scandit.Barcode.Symbology.CODE39, true);
settings.setSymbologyEnabled(Scandit.Barcode.Symbology.EAN8, true);
settings.setSymbologyEnabled(Scandit.Barcode.Symbology.QR, true);
settings.setSymbologyEnabled(Scandit.Barcode.Symbology.DATA_MATRIX, true);
settings.codeDuplicateFilter = 3500;
picker = new Scandit.BarcodePicker(settings);
picker.setMargins(new Scandit.Margins(0, 50, 0, 50), null, 0);
picker.continuousMode = true;
picker.orientations = Scandit.BarcodePicker.Orientation.PORTRAIT;
picker.getOverlayView().setGuiStyle(0);
picker.show($scope.scanSuccess, null, $scope.scanFailure);
picker.startScanning();
```

**Comment by moritz-scandit** [comment]
Hi,

Since you have sent this on both channels I will just answer you here instead of through our support mail as well.

I assume this is happening under Android? Under Android you can't set any margin for top/bottom as the OS itself will simply snap to fullscreen again if the margins are too small. Please try setting the margins to something like 0, 200, 0, 100 and see what happens. It will probably not be full screen again. Sadly we can't do anything about this as it's simply how Android acts.

If this is not what is happening please write what phone you are testing on.

**Comment by linkito** [comment]
Hi again moritz... The problem is when I set the both margins (top and bottom). If I set only a margin, top for example, all runs well (from 1 to 50)... But when I set the margin top  50, and bottom  50, or top  200 and bottom 200... the scanner opens in fullscreen... :(. I need play with the margins because in continuous mode, I change the margins to show a box area with the information of the scanned code.

The device that I'm testing is GT-I9301I

Thanks!

**Comment by linkito** [comment]
Now, I'm trying change the aspect of my app to play only with the top margin, because the issue of use the top and bottom margin I expect that not be solved in a time.
But with a top margin only I have issues too... This is the problem:

I start the scan with a setMargins with only the top margin.... all right here:
```javascript
// scannerBox.top is 43
picker.setMargins(new Scandit.Margins(0, parseInt(scannerBox.top), 0, 0), null, 0);
```

After, when a barcode is scanned, I change the margins with the same code... and all its ok again:
```javascript
// scannerBox.bottom is 260
picker.setMargins(new Scandit.Margins(0, parseInt(scannerBox.bottom), 0, 0), null, 0);
```
But after 5 seconds... I reset the margin to the first mode... and the scanner goes crazy... I have to relaunch the app to the margins works well:
```javascript
timeOutTicket = $timeout(function () {
   picker.setMargins(new Scandit.Margins(0, parseInt(scannerBox.top), 0, 0), null, 0);
}, 5000);
```
Here you have a video showing this issue:
[https://mega.nz/#!KABWlIRI!8lJMZ9NGmflldK1fcuiwCukYD8MfAyVE2VNnt3_FlSY](url)

**Comment by moritz-scandit** [comment]
I see, that looks like an odd issue that we have not seen so far. Thank you for providing so much information. We will have a look at this on Monday and get back to you asap.

**Comment by moritz-scandit** [comment]
I looked into this and the problem is what I mentioned in my previous post. The Android OS snaps the camera preview to the top under certain circumstances. If I change the code such that it goes from 260 to 200 it works fine but if I go to 120 it snaps to fullscreen again. I am not sure why it doesn't already do it at the beginning but I am also not sure why it happens at all. The same thing does not happen with a normal view, a normal view is resized exactly as you would expect. The sad fact is that the SurfaceView onto which the camera preview has to be rendered under the camera api 1 simply has problems with resizing. Camera api 2 might resolve this (I haven't actually tested it) but there are so few devices that properly support it that relying on it would be impossible.

Under native Android this whole thing isn't that big of a problem as you can simply put other views on top of the camera's surface view and animate them, however, as a phonegap plugin we can't do anything similar.

In conclusion we can sadly not help further with this as it is a limitation of the SurfaceView provided by Android.

**Comment by biasmv** [comment]
In 4.14, we have switched the implementation of the barcode picker on Android to use TextureView whenever possible. The snapping issue you were seeing should no longer happen. I'd suggest that you update your application to 4.14.

I'm going to close the ticket for now. In case you are having issues with 4.14 and snapping, feel free to reopen the ticket.

Best,
Marco
